frame/offences/src/migration.rs

```diff
@@ -0,0 +1,122 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2021 Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use super::Config;
+use frame_support::{storage::StorageValue, traits::Get, weights::Weight};
+use sp_staking::offence::OnOffenceHandler;
+
+mod deprecated {
+    use crate::{Config, OffenceDetails, Perbill, SessionIndex};
+    use frame_support::{sp_std::marker::PhantomData, storage::generator::StorageValue};
+    use sp_std::vec::Vec;
+
+    /// Type of data stored as a deferred offence
+    pub type DeferredOffenceOf<T> = (
+        Vec<
+            OffenceDetails<
+                <T as frame_system::Config>::AccountId,
+                <T as Config>::IdentificationTuple,
+            >,
+        >,
+        Vec<Perbill>,
+        SessionIndex,
+    );
+
+    /// Deferred reports that have been rejected by the offence handler and need to be submitted
+    /// at a later time.
+    pub struct DeferredOffences<T: Config>(PhantomData<T>);
+
+    impl<T: Config> StorageValue<Vec<DeferredOffenceOf<T>>> for DeferredOffences<T> {
+        type Query = Vec<DeferredOffenceOf<T>>;
+        fn module_prefix() -> &'static [u8] {
+            b"Offences"
+        }
+        fn storage_prefix() -> &'static [u8] {
+            b"DeferredOffences"
+        }
+        fn from_optional_value_to_query(v: Option<Vec<DeferredOffenceOf<T>>>) -> Self::Query {
+            v.unwrap_or_else(|| Default::default())
+        }
+        fn from_query_to_optional_value(v: Self::Query) -> Option<Vec<DeferredOffenceOf<T>>> {
+            Some(v)
+        }
+    }
+}
```

Comment:
The `generate_storage_alias!` helper doesn't work with generics 😢 
```rust
generate_storage_alias!(
    "Offences",
    DeferredOffences<T> => Value<Vec<DeferredOffenceOf<T>>>
);
```

So I implemented the StorageValue manually here, let me know if there's a cleaner way to implement it.

Reply:
There's a pretty clear example of this in the macro doc and should be supported: 

```
/// // generate a map from `Config::AccountId` (with hasher `Twox64Concat`) to `Vec<u8>`
/// trait Config { type AccountId: codec::FullCodec; }
/// generate_storage_alias!(
/// 	Prefix, GenericStorage<T: Config> => Map<(Twox64Concat, T::AccountId), Vec<u8>>
/// );
```

Reply:
Thx @kianenigma you are right! The only issue I have now is that it doesn't generates a `pub` storage type, so I can't use it outside `mod deprecated`.

```rust
   |     let deferred = <deprecated::DeferredOffences<T>>::take();
   |                                 ^^^^^^^^^^^^^^^^ private type alias
```

One way to solve it is extend this macro is by accepting an pattern like this:
``` rust
generate_storage_alias!(
    Offences,
    pub DeferredOffences<T: Config> => Value<Vec<DeferredOffenceOf<T>>>
);
```

But I think it is out of the scope of this PR, for now I'll remove the `deprecated` module :)